### PR TITLE
Introduce token revocation with binding expiry

### PR DESCRIPTION
**Pull request by pamodaaw**
**Description**
Access token binding can be defined for the OAuth applications. Once the binding is defined, a unique access token is issued per binding. Therefore there is no use of keeping the access token once the binding is expired. So that this PR adds the functionality to revoke the access token when the access token binding is expired. This feature will work only if it is enabled in the service provider.

**Approach**

A new property is added to the OAuth/OIDC inbound authentication configuration.
image
When this property is enabled only the tokens issued for the application will get revoked once the IDP session terminates.

For this feature to work, the application should have a binding type defined. We have two binding types, `cookie` and `sso-session` as of now. When the user logs out of the application, the new event handler added to listen to the SESSION_TERMINATE event will be triggered. This handler will revoke the access token of the token binding reference issued for the application. 

The following flow chart explains the behavior of this feature.

image

**Follow up actions**
Update the default.json of <IS_HOME>/repository/resources/conf/ folder to enable this event listener by default

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp [hunk 1]
```diff
@@ -95,7 +95,7 @@
         supportedIdTokenEncryptionMethods =
                 client.getSupportedIDTokenAlgorithms().getSupportedIdTokenEncryptionMethods();
         supportedTokenBindingsMetaData = client.getSupportedTokenBindingsMetaData();
-
+
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit.jsp:98]: remove this

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java [hunk 2]
```diff
@@ -2439,4 +2442,59 @@ private boolean isAccessTokenExpired(long issuedTimeInMillis, long validityPerio
...
+                            String tokenId = resultSet.getString("TOKEN_ID");
+
+                            String[] scope = OAuth2Util.buildScopeArray(tokenScope);
+                            accessTokenDO.setAccessToken(token);
+                            accessTokenDO.setConsumerKey(consumerKey);
+                            accessTokenDO.setScope(scope);
+                            accessTokenDO.setAuthzUser(user);
+                            accessTokenDO.setTenantID(tenantId);
+                            accessTokenDO.setRefreshToken(refreshToken);
+                            accessTokenDO.setTokenId(tokenId);
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java:2484]: Are we using all of these values which are set to the accessTokenDO

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java [hunk 3]
```diff
@@ -2439,4 +2442,59 @@ private boolean isAccessTokenExpired(long issuedTimeInMillis, long validityPerio
...
+                    (resultSet, i) -> {
+                        String token = resultSet.getString("ACCESS_TOKEN");
+                        AccessTokenDO accessTokenDO = new AccessTokenDO();
+                        if (tokenMap.containsKey(token)) {
+                            AccessTokenDO tokenObj = tokenMap.get(token);
+                            String[] previousScope = tokenObj.getScope();
+                            String[] newSope = new String[tokenObj.getScope().length + 1];
+                            System.arraycopy(previousScope, 0, newSope, 0, previousScope.length);
+                            newSope[previousScope.length] = resultSet.getString(2);
+                            tokenObj.setScope(newSope);
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java:2470]: We are not using the scope values

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+import static org.wso2.carbon.identity.application.authentication.framework.util.FrameworkConstants.COMMONAUTH_COOKIE;
+import static org.wso2.carbon.identity.application.authentication.framework.util.FrameworkConstants.RequestParams.TYPE;
+
+/**
+ * Event handler for token revocation during access token binding expiration.
+ */
+public class TokenBindingExpiryEventHandler extends AbstractEventHandler {
+
+    private static final Log log = LogFactory.getLog(TokenBindingExpiryEventHandler.class);
+    private static final String SSO_SESSION_BINDING = "sso-session";
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java:60]: Use a common constant for this and the SSOSessionBasedCookieBinder

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+    @Override
+    public void handleEvent(Event event) throws IdentityEventException {
+
+        if (log.isDebugEnabled()) {
+            log.debug(event.getEventName() + " event received to TokenBindingExpiryEventHandler.");
+        }
+
+        if (StringUtils.equals(event.getEventName(), IdentityEventConstants.EventName.SESSION_TERMINATE.name())) {
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java:69]:
```suggestion
        if (I!dentityEventConstants.EventName.SESSION_TERMINATE.name().equals(event.getEventName())) {
            return true;
        }
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+ * Event handler for token revocation during access token binding expiration.
+ */
+public class TokenBindingExpiryEventHandler extends AbstractEventHandler {
+
+    private static final Log log = LogFactory.getLog(TokenBindingExpiryEventHandler.class);
+    private static final String SSO_SESSION_BINDING = "sso-session";
+
+    @Override
+    public void handleEvent(Event event) throws IdentityEventException {
+
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java:64]: Don't we need to check is the listener enabled here?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        }
+
+        if (StringUtils.equals(event.getEventName(), IdentityEventConstants.EventName.SESSION_TERMINATE.name())) {
+
+            HttpServletRequest request = getHttpRequestFromEvent(event);
+            Map<String, Object> eventProperties = event.getEventProperties();
+            AuthenticationContext context = (AuthenticationContext) eventProperties.get(IdentityEventConstants
+                    .EventProperty.CONTEXT);
+            try {
+                if (StringUtils.equals(request.getParameter(TYPE), FrameworkConstants.RequestType.CLAIM_TYPE_OIDC)) {
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java:76]:
```suggestion
                if (FrameworkConstants.RequestType.CLAIM_TYPE_OIDC.equals(request.getParameter(TYPE))) {
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+
+                    String consumerKey = context.getRelyingParty();
+                    String bindingType = OAuth2Util.getAppInformationByClientId(consumerKey).getTokenBindingType();
+
+                    revokeTokensForBindingType(request, context.getLastAuthenticatedUser(), consumerKey, bindingType);
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java:81]: if there are no binding types we can skip this

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java (hunk 3 above)
Comment (pamodaaw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java:2470]: We need the scopes to invalidate the cache. https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1413/files#diff-4e2e9dcf5da24d01c2cbe7ee9a8fd87dR205

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java (hunk 6 above)
Comment (pamodaaw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java:64]: The listener is hit only if it is subscribed for the specific event from the identity-event.properties file.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java (hunk 2 above)
Comment (pamodaaw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java:2484]: We need the token, consumer key and scope.

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add-finish-ajaxprocessor.jsp [hunk 9]
```diff
@@ -79,6 +80,10 @@
         bypassClientCredentials = true;
     }
     
+    if (request.getParameter("revokeTokensWhenIDPSessionTerminated") != null){
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/add-finish-ajaxprocessor.jsp:83]: format this line

File: components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit-finish-ajaxprocessor.jsp [hunk 10]
```diff
@@ -82,6 +83,10 @@
         bypassClientCredentials = true;
     }
     
+    if (request.getParameter("revokeTokensWhenIDPSessionTerminated") != null){
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth.ui/src/main/resources/web/oauth/edit-finish-ajaxprocessor.jsp:86]: format this line

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Constants.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+ */
+public class OAuth2Constants {
+
+    /**
+     * Constants for access token binders.
+     */
+    public static class TokenBinderType {
+
+        public static final String SSO_SESSION_BASED_TOKEN_BINDER = "sso-session";
+
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Constants.java:32]: remove this empty line.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Constants.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.oauth2;
+
+/**
+ * This class contains the constants required by the OAuth2 components.
+ */
+public class OAuth2Constants {
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Constants.java:24]: Do we need this class?

Comment (pamodaaw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Constants.java:24]: There's no class for the constants. Or else we can define this constant in the `org.wso2.carbon.identity.oauth.common` component

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java (hunk 8 above)
Comment (pamodaaw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/bindings/handlers/TokenBindingExpiryEventHandler.java:84]: added a null check and called this method

**Pull request merged**